src/ucp/dt/datatype_iter.c

```diff
@@ -72,6 +72,10 @@ void ucp_datatype_iter_iov_mem_dereg(ucp_context_h context,
     ucp_mem_h *memh = dt_iter->type.iov.memh;
     size_t iov_index, length;
 
+    if (memh == NULL) {
+        return;
+    }
+
     ucp_datatype_iter_iov_for_each(iov_index, length, dt_iter) {
         ucp_memh_put(context, memh[iov_index], 0);
```

Comment:
Shouldn't  it be smth like this:

```C
void ucp_memh_dereg(ucp_context_h context, ucp_mem_h memh)
{
    if (memh != NULL) {
        ucp_memh_put(context, dt_iter->type.contig.memh, 0);
        memh = NULL;
    }
}

void ucp_datatype_iter_iov_mem_dereg(ucp_context_h context,
                                     ucp_datatype_iter_t *dt_iter)
{
    ucp_mem_h *memh = dt_iter->type.iov.memh;
    size_t iov_index, length;

    if (memh == NULL) {
        return;
    }

    ucp_datatype_iter_iov_for_each(iov_index, length, dt_iter) {
        ucp_memh_dereg(context, memh[iov_index]);
    }

    ucs_free(memh);
    dt_iter->type.iov.memh = NULL;
}

static UCS_F_ALWAYS_INLINE void
ucp_datatype_iter_mem_dereg(ucp_context_h context, ucp_datatype_iter_t *dt_iter,
                            unsigned dt_mask)
{
    if (ucp_datatype_iter_is_class(dt_iter, UCP_DATATYPE_CONTIG, dt_mask)) {
        ucp_memh_dereg(context, dt_iter->type.contig.memh);
    } else if (ucp_datatype_iter_is_class(dt_iter, UCP_DATATYPE_IOV, dt_mask)) {
        ucp_datatype_iter_iov_mem_dereg(context, dt_iter);
    }
}
```

Reply:
unification is nice, but do we really need to have a branch for checking every iov memh if dt is registered?

Reply:
no, this cleanup code is broken anyway, `dt_iter->memh` is set only on [line 65](https://github.com/openucx/ucx/blob/29fa51260b0012872582f8b6b3563994e324a082/src/ucp/dt/datatype_iter.c#L65)

Reply:
i'll address this issue in this PR  as well